### PR TITLE
feat(sql rollup): create sql vqb rollup step

### PR DESCRIPTION
**Pull request by raphaelvignes**
* Add Hierarchical Rollup Step for SQL VQB

### CODE REVIEW

**Comment by Sanix-Darker** [comment]
[[Snowflake SQL VQB] - Hierarchical rollup (5)](https://trello.com/c/P70lG6DY/1219-snowflake-sql-vqb-hierarchical-rollup-5)

File: server/tests/steps/sql_translator/test_rollup.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from weaverbird.backends.sql_translator.steps.rollup import translate_rollup
+from weaverbird.pipeline.steps import RollupStep
+
+
+@pytest.fixture
+def sql_query_describer():
+    return Mock(return_value={'toto': 'int', 'raichu': 'str'})
+
+
+def test_rollup(query, sql_query_describer):
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/test_rollup.py:15]:
```suggestion
def test_translate_rollup(query, sql_query_describer):
```

File: server/tests/steps/sql_translator/test_rollup.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        name='rollup',
+        hierarchy=['CONTINENT', 'COUNTRY', 'CITY'],
+        aggregations=[
+            {'newcolumns': ['VALUE'], 'aggfunction': 'sum', 'columns': ['VALUE']},
+        ],
+    )
+    result = translate_rollup(step, query, index=1, sql_query_describer=sql_query_describer)
+    assert (
+        result.transformed_query
+        == '''WITH SELECT_STEP_0 AS (SELECT * FROM products), ROLLUP_STEP_1 AS (SELECT CONTINENT, COUNTRY, CITY, COALESCE(CITY, COUNTRY, CONTINENT) AS LABEL, CASE WHEN CITY IS NOT NULL THEN 'CITY' WHEN COUNTRY IS NOT NULL THEN 'COUNTRY' WHEN CONTINENT IS NOT NULL THEN 'CONTINENT' ELSE '' END AS LEVEL, CASE WHEN LEVEL = 'COUNTRY' THEN "CONTINENT" WHEN LEVEL = 'CITY' THEN "COUNTRY" ELSE NULL END AS PARENT, SUM(VALUE) AS VALUE FROM SELECT_STEP_0 GROUP BY ROLLUP(CONTINENT, COUNTRY, CITY) HAVING CONTINENT IS NOT NULL)'''
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/test_rollup.py:26]: Can we have this in multiline for a good readability ?

File: server/tests/steps/sql_translator/test_rollup.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    )
+    assert result.query_name == 'ROLLUP_STEP_1'
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/test_rollup.py:28]: Can you add the assert test on `selection_query` ?
```python
assert query.selection_query == 'SELECT ........'
```

File: server/tests/steps/sql_translator/test_rollup.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            original_name='raichu',
+            type='STR',
+            original_type='str',
+            alias=None,
+            delete=False,
+        ),
+    }
+
+
+def test_rollup_groupby(query, sql_query_describer):
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/test_rollup.py:49]:
```suggestion
def test_translate_rollup_groupby(query, sql_query_describer):
```

File: server/tests/steps/sql_translator/utils/test_query_transformation.py [hunk 5]
```diff
@@ -44,6 +46,28 @@ def query():
...
+def aggregations(mocker):
+    return [
+        mocker.MagicMock(
+            columns=['ENRON', 'STANDARD OIL'],
+            agg_function='SUM',
+            new_columns=['SUM_ENRON', 'SUM_STANDARD_OIL'],
+        ),
+        mocker.MagicMock(columns=['ARAMCO'], agg_function='AVG', new_columns=['MOAAAR_OIL']),
+    ]
+
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/utils/test_query_transformation.py:59]: 👍🏾

File: server/tests/steps/sql_translator/utils/test_query_transformation.py [hunk 6]
```diff
@@ -44,6 +46,28 @@ def query():
...
+
+class FakeStep:
+    def __init__(self, hierarchy, aggregations, level_col, label_col, parent_label_col, groupby):
+        self.hierarchy = hierarchy
+        self.aggregations = aggregations
+        self.level_col = level_col
+        self.label_col = label_col
+        self.parent_label_col = parent_label_col
+        self.groupby = groupby
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/utils/test_query_transformation.py:68]: 👍🏾

File: server/tests/steps/sql_translator/utils/test_query_transformation.py [hunk 7]
```diff
@@ -416,3 +440,33 @@ def test_generate_query_by_keeping_granularity(query):
...
+        hierarchy=['EXXON', 'CHEVRON', 'TEXACO'],
+        aggregations=aggregations,
+        level_col='ROCKFELLER',
+        label_col='CFP',
+        parent_label_col='ANGLO-PERSIAN OIL COMPANY',
+        groupby=['SAUDI ARABIA', 'IRAN'],
+    )
+    assert (
+        build_hierarchical_columns_list(step)
+        == """EXXON, CHEVRON, TEXACO, SAUDI ARABIA, IRAN, COALESCE(TEXACO, CHEVRON, EXXON) AS CFP, CASE WHEN TEXACO IS NOT NULL THEN 'TEXACO' WHEN CHEVRON IS NOT NULL THEN 'CHEVRON' WHEN EXXON IS NOT NULL THEN 'EXXON' ELSE '' END AS ROCKFELLER, CASE WHEN ROCKFELLER = 'CHEVRON' THEN "EXXON" WHEN ROCKFELLER = 'TEXACO' THEN "CHEVRON" ELSE NULL END AS ANGLO-PERSIAN OIL COMPANY, SUM(ENRON) AS SUM_ENRON, SUM(STANDARD OIL) AS SUM_STANDARD_OIL, AVG(ARAMCO) AS MOAAAR_OIL"""
```
Comment (Sanix-Darker) [inline server/tests/steps/sql_translator/utils/test_query_transformation.py:471]: Can we have this in multilines too ?

**Review by Sanix-Darker — APPROVED** [review]
Thanks for this PR !
It looks good to me !
I left some comments for you !

File: server/tests/steps/sql_translator/test_rollup.py (hunk 1 above)
Comment (raphaelvignes) [reply server/tests/steps/sql_translator/test_rollup.py:15]: :ok:

File: server/tests/steps/sql_translator/test_rollup.py (hunk 2 above)
Comment (raphaelvignes) [reply server/tests/steps/sql_translator/test_rollup.py:26]: :ok:

File: server/tests/steps/sql_translator/test_rollup.py (hunk 4 above)
Comment (raphaelvignes) [reply server/tests/steps/sql_translator/test_rollup.py:49]: :ok:

File: server/tests/steps/sql_translator/utils/test_query_transformation.py (hunk 7 above)
Comment (raphaelvignes) [reply server/tests/steps/sql_translator/utils/test_query_transformation.py:471]: yep

**Pull request merged**